### ISSUE TITLE
Alerts: Examples and documentation for dynamic alerts

### ISSUE THREAD

**Issue by jgiardino** [patternfly/patternfly#854]
As a consumer of patternfly, I should be able to find examples and documentation that explains how to implement dynamic alerts in a way that is accessible to all users.

**NOTE** _This issue has been partially implemented, and also partially duplicated in #1375. What's left here is primarily related to how we represent actions/links that are included in the contents of a toast notification._

**NOTE** _The Alert component can be used for both static and dynamic alerts._ 

**Suggested Updates**
- Include an a11y documentation section for the dynamic alert example 
  - This section should explain when to use `<a>` vs `<button class="pf-c-button pf-m-link pf-m-inline>` elements. See notes below.
- Update the Upgrade example for Alerts: 
  - 3rd Toast notification example updates:
    - Update the message to something like: "The server configuration changed. Go to the [Servers](#) page to reload the server."
  - 4th Toast notification example updates: 
    - Replace the kebab in the last PF4 toast notification with a Close button. 
    - Update the message to something like: "Great job! This is really working out. Go to the [Servers](#) page to access available actions."
  - Include a statement about not including the kebab for accessibility reasons, with a link to the a11y documentation section for Toasts for more information. 

**Updates already addressed**
- ~~Create a specific example for dynamic alerts. See notes below about dynamic alerts.~~ see issue #1375 for this update
- ~~Remove elements/attributes that should only exist for dynamic alerts from the other examples~~ see issue #1375 for this update
- ~~Update the screen-reader-only text that identifies the type of alert to include "alert" (for both types of alerts, in all instances of alerts):~~ completed in #1226 
    ```
    <h4 class="pf-c-alert__title">
      <span class="pf-screen-reader">
        Success alert: 
      </span> Success notification title
    </h4>
    ```
  NOTE: the sr-only utility class is changing to `pf-screen-reader` in #1143, which is reflected here.

**Dynamic alerts**
The following are aspects that are unique to dynamic alerts, and should be reflected in the examples. 

- Only dynamic alerts include aria live attributes. Static alerts do not. See #1375 for more details.
- Dynamic alerts could include a link or action, as long as it's incorporated into the message in a way that users would know how to access that link/action when the dynamic alert disappears. This is to accommodate users who will not be able to access the actions in a dynamic alert. For example:
  - Do Not Do This:
    > The build is complete. [Download](#)
  - Do This Instead: 
     using `<button class="pf-c-button pf-m-link pf-m-inline>Download</button>`

    > The build is complete. Go to the Builds page to [download](#).
  - Or Do This:
     using `<a href="url">Builds</a>`

    > The build is complete. Go to the [Builds](#) page to download.

**Notes about alerts and screen readers**
- When an element with aria live attributes is included on a page, the contents that are inserted into that element are announced to the user, but focus remains unchanged (i.e. getting to the contents of the alert if actions are included would require explicit user action to do so).
- In JAWS and VO, clickable elements that included inside a live region are not announced as clickable elements. (i.e. there's no indication that the message includes something that the screen reader user could interact with). 
- In JAWS (but not VO), using a heading element (e.g. `<h3>`) results in the heading text getting announced as a heading

**Comment by matthewcarleton** [comment]
This sounds really good @jgiardino, thanks for putting this together. I'd say the second of the two suggested texts makes more sense in the context of the message, but I'm not sure if we want to/can be that prescriptive.

**Comment by jgiardino** [comment]
I updated the description to include "alert" as part of the text that's announced to screen reader users. But I'm not sure if this is a change we should make to the Alert component as a whole, or just the dynamic alerts (i.e. toast notifications).

For example if the page loads with a static warning alert already visible is it still an alert?

  "Warning alert: A newer version is available."
or
  "Warning: A newer version is available."
(the visible text for either case would only be "A newer version is available."

@stacymcauliffe I'd love any input from you on this one :-)

**Comment by stacymcauliffe** [comment]
@jgiardino I think adding in both places is fine since they're all alerts, simply being displayed to the user in a different way.

**Comment by jgiardino** [comment]
Thanks, @stacymcauliffe! I updated the description to clarify that this text change is for all alerts.

**Comment by andybraren** [comment]
@jgiardino @stacymcauliffe What do we think about changing the word "alert" to the more colloquial "notification" when the alert is included within a Toast Alert Group in the top-right corner?

Inline Alert: "Warning alert: The inline thing you were just doing failed"
Toast Alert: "Warning notification: The server is on fire please help."

**Comment by stacymcauliffe** [comment]
Notification def feels less alarming than "alert," and I think we generally reserve toast notifications for stuff that isn't essential, so I would be fine with this change.

Also, I really like this example text. I'd just hope that the server fire would prompt an alert rather than a laid back notification.

**Comment by jgiardino** [comment]
I could go either way on this. I kind of lean toward using one term for both if just for simplicity, but don't have a strong preference for either term.

Also, based on @stacymcauliffe's feedback it sounds like these cases would be flipped from what @andybraren used if we used different terms for inline vs toast?

Inline Alert: "Warning notification: The inline thing you were just doing failed"
Toast Alert: "Warning alert: The server is on fire please help."

Also, just a point of clarification to make it more confusing. Toast alerts are always dynamic (i'm pretty sure), but inline alerts could be either static (visible on page load) or dynamic (inserted after page load based on user action or some other cause).

**Comment by mcoker** [comment]
cc @jessiehuff

**Comment by jessiehuff** [comment]
It looks like some of these points are still valid and that Jenn updated the description with the aspects still outstanding, mostly related to our examples and documentation. I think we can keep this open.
